### PR TITLE
conn.go: default close handler should not return ErrCloseSent.

### PR DESCRIPTION
**Pull request by pnx**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Go Version Update
- [ ] Dependency Update

## Description

Noticed a change in default close handler from 1.5.0 to 1.5.1

1.5.1 Now returns the error from `WriteControl` with `CloseMessage` type. However this results in ErrCloseSent returned if the connection initiated the close handshake. This is normally correct as the connection should not be able to write after a close handshake is initiated, except when calling the close handler. in that case nil should be returned so that `advanceFrame()` can return the actual close message.

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Run verifications and test

- [ ] `make verify` is passing (fails with `GO-2023-2186 Incorrect detection of reserved device names on Windows in path/filepath` that is unrelated to this change)
- [x] `make test` is passing

### CODE REVIEW

**Comment by pennystevens** [comment]
A similar change is required in the ping handler.  Consider this scenario:

1.  PING message is in flight from peer
2. the application writes a CLOSE message
3. the peer echos the CLOSE message
4. the PING message is received, connection responds with PONG
5. the CLOSE message is received, connection echos the CLOSE message.

This PR addresses step 5 where ErrCloseSent is returned instead of the CLOSE message.

Application never gets to step 5 because the PING handler returns ErrCloseSent and this error is returned from NextReader, ReadMessage, etc. 

In summary: When the connection automatically sends a message in response to a received message, ErrCloseSent should be ignored.  

Version 1.5.0 also ignored timeout errors in this scenario.

**Comment by pnx** [comment]
@coreydaley can you take a look at this?

**Review by coreydaley — APPROVED** [review]
